Sources/Tracking/Background Queue/QueueStorage.swift

```diff
@@ -0,0 +1,82 @@
+import Foundation
+
+public protocol QueueStorage: AutoMockable {
+    func getInventory() -> [QueueTaskItem]
+    func saveInventory(_ inventory: [QueueTaskItem]) -> Bool
+
+    func create(type: QueueTaskType, data: Data) -> Bool
+    func update(storageId: String, runResults: QueueTaskRunResults) -> Bool
+    func get(storageId: String) -> QueueTask?
+}
+
+// sourcery: InjectRegister = "QueueStorage"
+public class FileManagerQueueStorage: QueueStorage {
+    private let fileStorage: FileStorage
+    private let jsonAdapter: JsonAdapter
+
+    init(fileStorage: FileStorage, jsonAdapter: JsonAdapter) {
+        self.fileStorage = fileStorage
+        self.jsonAdapter = jsonAdapter
+    }
+
+    public func getInventory() -> [QueueTaskItem] {
+        guard let data = fileStorage.get(type: .queueInventory, fileId: nil) else { return [] }
+
+        let inventory: [QueueTaskItem] = jsonAdapter.fromJson(data, decoder: nil) ?? []
+
+        return inventory
+    }
+
+    public func saveInventory(_ inventory: [QueueTaskItem]) -> Bool {
+        guard let data = jsonAdapter.toJson(inventory, encoder: nil) else {
+            return false
+        }
+
+        return fileStorage.save(type: .queueInventory, contents: data, fileId: nil)
+    }
+
+    public func create(type: QueueTaskType, data: Data) -> Bool {
+        let newTaskStorageId = UUID().uuidString
+        let newQueueTask = QueueTask(storageId: newTaskStorageId, type: type, data: data,
+                                     runResults: QueueTaskRunResults(totalRuns: 0))
+
+        if !update(queueTask: newQueueTask) {
```

Comment:
If I understand correctly, you're asking when would `update()` ever be `false` here with creating a brand new queue item? 

The answer is that: we are saving data to the device's file system and because of that, we *always* run the risk of this function being `false`. The biggest reason for a save failing against a file system is the iOS device runs out of storage but there are other scenarios. 

The functions in `FileStorage` all use bool returns indicating if the file system write was successful or not for that reason - saving to a file system can always fail.

Reply:
gotcha. feels pretty unexpressive to use a bool instead of an error type, but seems fine for now. I would imagine for debug logging at some point we may want more info than just a false, can leave to then

Reply:
Debugging is very important. I tried to [centralize the logging to 1 place](https://github.com/customerio/customerio-ios/blob/7dd6cf564f95d57736a7ffbda1d881f95cfa8663/Sources/Tracking/Util/FileStorage.swift#L80) when there is an error in the code instead of N.